From bdf02a622db2d6b4c8569eabfa7617c28d09079e Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:11:59 -0400
Subject: [powerpc] powernv: Reset root port in firmware

Message-id: <1410545655-205645-491-git-send-email-dzickus@redhat.com>
Patchwork-id: 94412
O-Subject: [RHEL7 PATCH 490/626] powerpc/powernv: Reset root port in firmware
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit fd5cee7ce8f488768f918e73231d4859a520eb33
Author: Gavin Shan <gwshan@linux.vnet.ibm.com>
Date:   Thu Apr 24 18:00:22 2014 +1000

    powerpc/powernv: Reset root port in firmware

    Resetting root port has more stuff to do than that for PCIe switch
    ports and we should have resetting root port done in firmware instead
    of the kernel itself. The problem was introduced by commit 5b2e198e
    ("powerpc/powernv: Rework EEH reset").

    Cc: linux-stable <stable@vger.kernel.org>
    Signed-off-by: Gavin Shan <gwshan@linux.vnet.ibm.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/platforms/powernv/eeh-ioda.c b/arch/powerpc/platforms/powernv/eeh-ioda.c
index 3b76e58..697d620 100644
--- a/arch/powerpc/platforms/powernv/eeh-ioda.c
+++ b/arch/powerpc/platforms/powernv/eeh-ioda.c
@@ -527,12 +527,10 @@ static int ioda_eeh_reset(struct eeh_pe *pe, int option)
  int ret;
 
  /*
-  * The rules applied to reset, either fundamental or hot reset:
-  *
-  * We always reset the direct upstream bridge of the PE. If the
-  * direct upstream bridge isn't root bridge, we always take hot
-  * reset no matter what option (fundamental or hot) is. Otherwise,
-  * we should do the reset according to the required option.
+  * For PHB reset, we always have complete reset. For those PEs whose
+  * primary bus derived from root complex (root bus) or root port
+  * (usually bus#1), we apply hot or fundamental reset on the root port.
+  * For other PEs, we always have hot reset on the PE primary bus.
   *
   * Here, we have different design to pHyp, which always clear the
   * frozen state during PE reset. However, the good idea here from
@@ -546,7 +544,8 @@ static int ioda_eeh_reset(struct eeh_pe *pe, int option)
   ret = ioda_eeh_phb_reset(hose, option);
  } else {
   bus = eeh_pe_bus_get(pe);
-  if (pci_is_root_bus(bus))
+  if (pci_is_root_bus(bus) ||
+      pci_is_root_bus(bus->parent))
    ret = ioda_eeh_root_reset(hose, option);
   else
    ret = ioda_eeh_bridge_reset(bus->self, option);
-- 
1.7.1